pkg/liquidity-source/uniswap-v1/pool_simulator.go

```diff
@@ -0,0 +1,207 @@
+		Gas:            s.gas.Swap,
+	}, nil
+}
+
+func (s *PoolSimulator) CalcAmountIn(param poolpkg.CalcAmountInParams) (*poolpkg.CalcAmountInResult, error) {
+	var (
+		tokenAmountOut = param.TokenAmountOut
+		tokenIn        = param.TokenIn
+	)
+	indexIn, indexOut := s.GetTokenIndex(tokenIn), s.GetTokenIndex(tokenAmountOut.Token)
+	if indexIn < 0 || indexOut < 0 {
+		return nil, ErrInvalidToken
+	}
+
+	amountOut, overflow := uint256.FromBig(tokenAmountOut.Amount)
+	if overflow {
+		return nil, ErrInvalidAmountOut
+	}
+
+	if amountOut.Cmp(number.Zero) <= 0 {
+		return nil, ErrInsufficientOutputAmount
+	}
+
+	reserveIn, overflow := uint256.FromBig(s.Pool.Info.Reserves[indexIn])
+	if overflow {
+		return nil, ErrInvalidReserve
+	}
+
+	reserveOut, overflow := uint256.FromBig(s.Pool.Info.Reserves[indexOut])
+	if overflow {
+		return nil, ErrInvalidReserve
+	}
+
+	amountIn, err := s.getOutputPrice(amountOut, reserveIn, reserveOut)
+	if err != nil {
+		return nil, err
+	}
+
+	if amountIn.Cmp(reserveIn) > 0 {
+		return nil, ErrInsufficientLiquidity
+	}
+
+	return &poolpkg.CalcAmountInResult{
+		TokenAmountIn: &poolpkg.TokenAmount{Token: s.Pool.Info.Tokens[indexOut], Amount: amountIn.ToBig()},
+		Fee:           &poolpkg.TokenAmount{Token: s.Pool.Info.Tokens[indexIn], Amount: integer.Zero()},
+		Gas:           s.gas.Swap,
+	}, nil
+}
+
+func (s *PoolSimulator) UpdateBalance(params poolpkg.UpdateBalanceParams) {
+	indexIn, indexOut := s.GetTokenIndex(params.TokenAmountIn.Token), s.GetTokenIndex(params.TokenAmountOut.Token)
+	if indexIn < 0 || indexOut < 0 {
+		return
+	}
+
+	s.Pool.Info.Reserves[indexIn] = new(big.Int).Add(s.Pool.Info.Reserves[indexIn], params.TokenAmountIn.Amount)
+	s.Pool.Info.Reserves[indexOut] = new(big.Int).Sub(s.Pool.Info.Reserves[indexOut], params.TokenAmountOut.Amount)
+}
+
+func (s *PoolSimulator) GetMetaInfo(_ string, _ string) interface{} {
+	return PoolMeta{
+		BlockNumber: s.Pool.Info.BlockNumber,
+	}
+}
+
+// def getInputPrice(input_amount: uint256, input_reserve: uint256, output_reserve: uint256) -> uint256:
+//
+//	assert input_reserve > 0 and output_reserve > 0
+//	input_amount_with_fee: uint256 = input_amount * 997
+//	numerator: uint256 = input_amount_with_fee * output_reserve
+//	denominator: uint256 = (input_reserve * 1000) + input_amount_with_fee
+//	return numerator / denominator
+func (s *PoolSimulator) getInputPrice(inputAmount, inputReserve, outputReserve *uint256.Int) (*uint256.Int, error) {
+	if inputReserve.CmpUint64(0) <= 0 || outputReserve.CmpUint64(0) <= 0 {
+		return nil, ErrInsufficientLiquidity
+	}
+
+	inputAmountWithFee := new(uint256.Int).Mul(inputAmount, uint256.NewInt(997))
+	numerator := new(uint256.Int).Mul(inputAmountWithFee, outputReserve)
+	denominator := new(uint256.Int).Add(
+		new(uint256.Int).Mul(
+			inputReserve,
+			uint256.NewInt(1000),
+		),
+		inputAmountWithFee,
+	)
+
+	return new(uint256.Int).Div(numerator, denominator), nil
```

Comment:
hey you can refer to this comment for some optimization tips. they are trivial, but I think it's worth knowing about:

https://github.com/KyberNetwork/kyberswap-dex-lib/pull/551#discussion_r1810816801

Reply:
for example in this case

```suggestion
	inputAmountWithFee := new(uint256.Int).Mul(inputAmount, U997)
	numerator := inputAmountWithFee.Mul(inputAmountWithFee, outputReserve)
	
	var denominator uint256.Int
	denominator.Add(
		denominator.Mul(
			inputReserve,
			U1000,
		),
		inputAmountWithFee,
	)

	return numerator.Div(numerator, &denominator), nil
```

1 heap + 1 stack alloc (+2 reused heap alloc for the constants) instead of 7 heap allocs

Reply:
note that these optimizations are more error-prone and harder to read, so should be used for small manageable code blocks and be tested sufficiently

Reply:
https://github.com/KyberNetwork/kyberswap-dex-lib/pull/552/commits/2aa5865c7e544d1bf58b073c53e24b9ec2259349